### PR TITLE
Template - Replication

### PR DESCRIPTION
**Pull request by Or-Geva**

### CODE REVIEW

File: artifactory/commands/replication/create.go [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+		}
+		return servicesManager.CreatePullReplication(*params)
+	case jobType == Push:
+		params, err := getPushParams(jsonConfig)
+		if err != nil {
+			return err
+		}
+		return servicesManager.CreatePushReplication(*params)
+	}
+	return errorutils.CheckError(errors.New("Cloud not detect replication job type "))
```
Comment (eyalbe4) [inline artifactory/commands/replication/create.go:96]: Redundant white-space at the end of the message.

File: artifactory/commands/replication/create.go [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+}
+
+func getRepoKey(jsonConfig []byte) (string, error) {
```
Comment (eyalbe4) [inline artifactory/commands/replication/create.go:99]: This function seems to be unused.

File: artifactory/commands/replication/show.go [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+func (rsc *ReplicationShowCommand) CommandName() string {
+	return "rt_replication_show"
+}
+
+func (rsc *ReplicationShowCommand) Run() error {
+	servicesManager, err := rtUtils.CreateServiceManager(rsc.rtDetails, false)
+	if err != nil {
+		return err
+	}
+	rsc.showResult, err = servicesManager.ShowReplication(rsc.repoKey)
```
Comment (eyalbe4) [inline artifactory/commands/replication/show.go:42]: The API invoked should be **servicesManager.GetReplication(rsc.repoKey)**. Then, the struct received should be indented and printed using log.Output.

File: artifactory/commands/replication/template.go [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+	},
+	RepoKey: {
+		Msg:          "Enter repo key",
+		PromptPrefix: ">",
+		AllowVars:    true,
+		Writer:       utils.WriteStringAnswer,
+		MapKey:       RepoKey,
+		Callback:     nil,
+	},
+	Username: {
```
Comment (eyalbe4) [inline artifactory/commands/replication/template.go:120]: We should not ask for a server URL, username and password, but for a server ID instead, The template should include the server ID only.

File: artifactory/commands/replication/template.go [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+
+func getAllPossibleOptionalRepoConfKeys(values ...string) []prompt.Suggest {
+	optionalKeys := []string{utils.WriteAndExist, Enabled, SyncDeletes, SyncProperties, SyncStatistics, PathPrefix, EnableEventReplication, SocketTimeoutMillis}
+	if len(values) > 0 {
+		optionalKeys = append(optionalKeys, values...)
+	}
+	return utils.GetSuggestsFromKeys(optionalKeys, suggestionMap)
+}
+
+//duplicate code
```
Comment (eyalbe4) [inline artifactory/commands/replication/template.go:194]: What this comment mean?

File: artifactory/cli.go [hunk 6]
```diff
@@ -662,6 +667,58 @@ func GetCommands() []cli.Command {
...
+			HelpName:     common.CreateUsage("rt rtc", replicationtemplate.Description, replicationtemplate.Usage),
+			UsageText:    replicationtemplate.Arguments,
+			ArgsUsage:    common.CreateEnvVars(),
+			BashComplete: common.CreateBashCompletionFunc(),
+			Action: func(c *cli.Context) error {
+				return replicationTemplateCmd(c)
+			},
+		},
+		{
+			Name:         "replication-job-create",
```
Comment (eyalbe4) [inline artifactory/cli.go:684]: replication-job-create --> replication-create
(let's change the short name also and the help.go file)

File: artifactory/cli.go [hunk 7]
```diff
@@ -662,6 +667,58 @@ func GetCommands() []cli.Command {
...
+			HelpName:     common.CreateUsage("rt rjc", replicationjobcreate.Description, replicationjobcreate.Usage),
+			UsageText:    replicationjobcreate.Arguments,
+			ArgsUsage:    common.CreateEnvVars(),
+			BashComplete: common.CreateBashCompletionFunc(),
+			Action: func(c *cli.Context) error {
+				return replicationCreateCmd(c)
+			},
+		},
+		{
+			Name:         "replication-job-delete",
```
Comment (eyalbe4) [inline artifactory/cli.go:697]: replication-job-delete --> replication-delete

File: artifactory/cli.go [hunk 8]
```diff
@@ -662,6 +667,58 @@ func GetCommands() []cli.Command {
...
+			Aliases:      []string{"rjd"},
```
Comment (eyalbe4) [inline artifactory/cli.go:698]: rjd --> rdel

File: artifactory/cli.go [hunk 9]
```diff
@@ -662,6 +667,58 @@ func GetCommands() []cli.Command {
...
+			HelpName:     common.CreateUsage("rt rjd", replicationdelete.Description, replicationdelete.Usage),
+			UsageText:    replicationdelete.Arguments,
+			ArgsUsage:    common.CreateEnvVars(),
+			BashComplete: common.CreateBashCompletionFunc(),
+			Action: func(c *cli.Context) error {
+				return replicationDeleteCmd(c)
+			},
+		},
+		{
+			Name:         "replication-job-show",
```
Comment (eyalbe4) [inline artifactory/cli.go:710]: replication-job-show --> replication-show
(let's change the short name also)

File: docs/artifactory/replicationjobdelete/help.go [hunk 10]
```diff
@@ -0,0 +1,8 @@
+package replicationdelete
+
+const Description = "Remove a replication repository job from Artifactory."
+
+var Usage = []string{`jfrog rt rjd <repository key>`}
+
+const Arguments string = `	repository key
+		The repository for which the replication configuration will be delete.`
```
Comment (eyalbe4) [inline docs/artifactory/replicationjobdelete/help.go:8]: The repository for which the replication configuration will be delete.` -->
The repository from which the replication configuration will be deleted.`

File: docs/artifactory/replicationjobshow/help.go [hunk 11]
```diff
@@ -0,0 +1,8 @@
+package replicationshow
+
+const Description = "Show a replication repository job configuration from Artifactory."
+
+var Usage = []string{`jfrog rt rjs <repository key>`}
+
+const Arguments string = `	repository key
+		The repository for which the data will be display.`
```
Comment (eyalbe4) [inline docs/artifactory/replicationjobshow/help.go:8]: ...display --> ...displayed

File: docs/artifactory/replicationtemplate/help.go [hunk 12]
```diff
@@ -0,0 +1,8 @@
+package replicationtemplate
+
+const Description = "Create a JSON template for repository creation or update."
```
Comment (eyalbe4) [inline docs/artifactory/replicationtemplate/help.go:3]: Incorrect description.

File: docs/artifactory/replicationtemplate/help.go [hunk 13]
```diff
@@ -0,0 +1,8 @@
...
+
+var Usage = []string{`jfrog rt rtc <template path>`}
```
Comment (eyalbe4) [inline docs/artifactory/replicationtemplate/help.go:5]: Isn't "jfrog rt rtc" already taken for repo template?

File: docs/artifactory/replicationtemplate/help.go [hunk 14]
```diff
@@ -0,0 +1,8 @@
...
+
+const Arguments string = `	template path
+		Specifies the local file system path for the template file.`
```
Comment (eyalbe4) [inline docs/artifactory/replicationtemplate/help.go:8]: @liron-shalom is adding a suffix to this command, with the command that creates the template. Let's add the same here.

**Review by eyalbe4 — CHANGES_REQUESTED** [review]

**Review by eyalbe4 — CHANGES_REQUESTED** [review]
We should probably also add tests.

File: artifactory/commands/replication/show.go (hunk 3 above)
Comment (Or-Geva) [reply artifactory/commands/replication/show.go:42]: Fixed. As a result, get will split to `get pull` and `get push`

File: artifactory/commands/replication/template.go (hunk 5 above)
Comment (Or-Geva) [reply artifactory/commands/replication/template.go:194]: repo and replication use this code(duplicated) need to merge it carefully with  liron

File: docs/artifactory/replicationtemplate/help.go (hunk 13 above)
Comment (Or-Geva) [reply docs/artifactory/replicationtemplate/help.go:5]: No, for repo template  that is `rpt`.

File: docs/artifactory/replicationtemplate/help.go (hunk 14 above)
Comment (Or-Geva) [reply docs/artifactory/replicationtemplate/help.go:8]: Done

File: artifactory/commands/replication/template.go (hunk 4 above)
Comment (Or-Geva) [reply artifactory/commands/replication/template.go:120]: Done

**Pull request closed without merging**